src/ray/object_manager/object_manager.cc

```diff
@@ -655,4 +666,47 @@ void ObjectManager::ExecuteReceiveObject(const ClientID &client_id,
                  << "/" << config_.max_receives;
 }
 
+void ObjectManager::ReceiveFreeRequest(std::shared_ptr<TcpClientConnection> &conn,
+                                       const uint8_t *message) {
+  auto fr = flatbuffers::GetRoot<object_manager_protocol::FreeRequestMessage>(message);
+  std::vector<ObjectID> object_ids = from_flatbuf(*fr->object_ids());
+  // This RPC should come from another Object Manager.
+  // Do not spread this request.
+  bool spread = false;
+  FreeObjects(object_ids, spread);
+}
+
+void ObjectManager::FreeObjects(const std::vector<ObjectID> &object_ids, bool spread) {
+  buffer_pool_.FreeObjects(object_ids);
+  if (spread) {
+    SpreadFreeObjectRequest(object_ids);
+  }
+}
+
+void ObjectManager::SpreadFreeObjectRequest(const std::vector<ObjectID> &object_ids) {
+  // This code path should be called from node manager.
+  flatbuffers::FlatBufferBuilder fbb;
+  flatbuffers::Offset<object_manager_protocol::FreeRequestMessage> request =
+      object_manager_protocol::CreateFreeRequestMessage(fbb, to_flatbuf(fbb, object_ids));
+  fbb.Finish(request);
+  auto callback = [this, &fbb](const RemoteConnectionInfo &connection_info) {
+    ray::Status status;
+    std::shared_ptr<SenderConnection> conn;
+    status = connection_pool_.GetSender(ConnectionPool::ConnectionType::MESSAGE,
+                                        connection_info.client_id, &conn);
+    RAY_CHECK_OK(status);
+    if (conn == nullptr) {
+      conn = CreateSenderConnection(ConnectionPool::ConnectionType::MESSAGE,
+                                    connection_info);
+      connection_pool_.RegisterSender(ConnectionPool::ConnectionType::MESSAGE,
+                                      connection_info.client_id, conn);
+    }
+    RAY_CHECK_OK(conn->WriteMessage(
+        static_cast<int64_t>(object_manager_protocol::MessageType::FreeRequest),
+        fbb.GetSize(), fbb.GetBufferPointer()));
+    RAY_CHECK_OK(
+        connection_pool_.ReleaseSender(ConnectionPool::ConnectionType::MESSAGE, conn));
+  };
+  RAY_CHECK_OK(object_directory_->GetAllClientInfo(callback));
```

Comment:
From the signature, it's not clear that `callback` will be run instantaneously here (as opposed to at some point in the future (e.g., in response to a reply from Redis), so it's not obvious that `fbb` will still be in scope and usable inside of `callback`.

Do you know what I mean? To judge correctness here I have to know about the implementation of `GetAllClientInfo`.

Actually, `callback` doesn't really seem like the right terminology since it runs instantaneously.